site_positions and precision bugs fixed

nitrobino opened the pull request:
Proposal to fix two erros:

The first error happens when you try to create an hdf5 file for the twisted bilayer, for example. An example of the unit cell is plotted below:

![Screenshot 2025-01-31 at 15 54 49](https://github.com/user-attachments/assets/8e7d1ccb-4f60-4b5b-806f-07cf98204ad3)

I have removed the hoppings for better viewing. 

From my understanding, since the creation of this bilayer was used from the HP module which generates this supercell from two graphene sheets with two sublattices each, Quantica records this lattice as having four sublattices: Ab, Bb; At; Bt (Each plotted with a different color). The system, however, has more than one site per unit cell. Therefore, the number of sublattices in Quantica does not necessarily equal the number of sites per unit cell. This will be important later.

When you try creating the hdf5 file for this system, an error happens:

<img width="757" alt="Screenshot 2025-01-31 at 17 04 07" src="https://github.com/user-attachments/assets/d3f9e4f8-d782-421c-9364-4e1b6b959051" />

The error comes from the function site_positions which is defined as:

<img width="566" alt="Screenshot 2025-01-31 at 16 05 11" src="https://github.com/user-attachments/assets/455096c8-a6ef-43af-9fc5-9861f5b9b649" />


The purpose of this function is , I think, to generate a matrix with the coordinates in real space of every site inside the unit cell, with a degenerancy for the number of atomic orbitals. The matrix position_atoms is therefore a matrix of shape (Number of Atoms in unit cell,Dimension of problem). Now, the reason I think this error is occuring is because this Quantica.norbitals function seems to return an array of size = The number of sublattices in the system (in this case, 4), with each entry being the number of atomic orbitals per site of that sublattice. Because i is running on the number of rows of position_atoms, wich is of length + Number of Atoms in unit Cell, num_orbitals[i] is bound to go out of bounds eventually. In my understanding, this function only works if the number of sites per unit cell equals the number of sublattices in Quantica.

I propose the following change: 

<img width="1329" alt="Screenshot 2025-01-31 at 17 18 47" src="https://github.com/user-attachments/assets/8804f42b-379f-4351-91ed-ddd4d1df723f" />


This way, sublat_ind is ensured to run over the number of sublattices of the system, the inner cicle does not go out of bounds and the final positions array has the desired outcome of being equal to the position_atoms matrix with repeated entries for the number of atomic orbitals in the system. With this modification, the program no longer gives an error and the hdf5 file is produced. 


The second error has to do with the precision for complex numbers, which is written in the code as 

![Screenshot 2025-01-28 at 15 06 05](https://github.com/user-attachments/assets/005ee1ed-14ef-4d02-ae1f-fb6596cfb177)

I believe the first precision == 1 is a typo and changed it to precision == 0.

Commit:
site_positions and precision bugs fixed

fernandopenaranda commented:
Dear @nitrobino thank you for your efforts in using and debugging this code and apologies for the delay in my response. Your PR is more than welcome. Allowing for more than one site per unit cell is crucial for QuantiKite. Your PR fixes satisfactory this problem when generating the hdf5 file. There is, however, a small issue within the propose code. I believe a for loop closure is required in the 186th line of kite_wrapper.jl. In order to accept the PR, could you please fix it?

Commit:
fixed size_positions and Norb

Commit:
added get_num_orbitals function that calculates total number of degrees of freedom in unit cell

fernandopenaranda commented:
Wonderful João. Tests passed and PR accepted! Thank you!

nitrobino commented:
Dear @fernandopenaranda,

Thank you for noticing that the previous code was indeed wrong. The way it was written, site_positions would return an array that would be N_Sublatt times larger than the correct array because of the for loop on sublat_ind. I believe I have fixed this error with the following code:

<img width="551" alt="Screenshot 2025-02-06 at 15 40 04" src="https://github.com/user-attachments/assets/76dcb624-ba4b-4177-a798-3edf46d74991" />

This code works under the assumption that the sites in h.lattice.unitcell.sites are organized by sublattices, i.e, the first sites pertain to the first sublattice, the second chunk of sites to the second, etc. I have verified this assumption for the case of the twisted bilayer model, therefore I assume this is how it works in every Quantica.jl lattice. If this assumption is not correct, please tell me. 

I have also noticed another problem in the NOrb dataset of the produced hdf5 file. In the above kite_wrapper.jl, it was set as sum(num_orbitals(h)), which will give the number of sublattices of the system (For the example shown in my first message, this number is equal to 4). I have changed this to be 

<img width="799" alt="Screenshot 2025-02-06 at 15 40 22" src="https://github.com/user-attachments/assets/b79774f9-fb91-4bf9-a048-1117dc753715" />

Where the function get_num_orbitals(h) is defined as 

<img width="617" alt="Screenshot 2025-02-06 at 15 42 40" src="https://github.com/user-attachments/assets/97ca1633-684b-479f-a172-d1e4b9d73390" />

This function calculates the total number of degrees of freedom in the unit cell. Both fields subsizes and blocksizes are vectors of size equal to the number of sublattices (4), where blocksizes has the number of atomic orbitals per atom in each sublattice and subsizes has the number of sites in each sublattice. 

With this changes, I have verified that the hdf5 file is produced and Kite sucessfully runs with that hdf5.

Please, feel free to check if this code works and please tell me if my assumption is correct!

Cheers!!

João Pinho

nitrobino commented:
@fernandopenaranda ,

I am very impressed with your speed in testing the PR, I couldn't even finish the above message on time! Glad the PR passed the tests and glad it could be of any help! Cheers!

fernandopenaranda commented:
Haha thanks for writing it down. I was following the changes you were doing in real time. Well done! Everything looks fine. I'll let you know in case I find any conflict when passing the hdf5 files to Kite. 
Cheers!

pablosanjose commented:
Bravo guys! Thank you
